src/common/snippets/src/pass/lowered/loop_init.cpp

```diff
@@ -0,0 +1,234 @@
+        const auto& out_tds = expr->get_outputs();
+        const auto& src_tensor = expr->get_inputs().front()->get_tensor();
+        const auto& dst_layout = out_tds[0]->get_layout();
+        const auto& dst_dim = *(dst_layout.rbegin() + dim_idx);
+        int64_t ptr_increment = 0;
+        // If relevant dim is not broadcasted, then ptr_increment is the dim stride in the new layout
+        if (!(src_tensor[dst_dim] == 1 && max_relevant_dim_size != 1))
+            ptr_increment = get_dim_stride(dst_dim, loop_layout, src_tensor);
+        ptr_increments.push_back(ptr_increment);
+    }
+    // Note: Le already accounted for loop_input vs inside loops layout mismatch. So we need non-dense output
+    // ptr_increments only if loop_input_layout doesn't match loop_output_layout
+    for (const auto& expr : loop_out_exprs) {
+        const auto& out_tds = expr->get_outputs();
+        const auto& dst_layout = out_tds[0]->get_layout();
+        const auto& dst_tensor = out_tds[0]->get_tensor();
+        const auto& dst_dim = *(loop_layout.rbegin() + dim_idx);
+        int64_t ptr_increment = 0;
+        // If relevant dim is not broadcasted, then ptr_increment is the dim stride in the new layout
+        if (!(dst_tensor[dst_dim] == 1 && max_relevant_dim_size != 1))
+            ptr_increment = get_dim_stride(dst_dim, dst_layout, dst_tensor);
+        ptr_increments.push_back(ptr_increment);
+    }
+
+    return ptr_increments;
+}
+
+std::vector<int64_t> LoopInit::init_finalization_offsets(const std::vector<int64_t>& ptr_increments, size_t work_amount) const {
+    std::vector<int64_t> finalization_offsets;
+    for (const auto& ptr_incr : ptr_increments) {
+        int64_t offset = -1 * ptr_incr * work_amount;
+        finalization_offsets.push_back(offset);
+    }
+    return finalization_offsets;
+}
+
+bool LoopInit::insertion(LoweredExprIR& linear_ir, const LoweredExprIR::LoweredLoopManager::LoweredLoopInfoPtr& loop_info,
+                         size_t loop_id, size_t dim_idx, bool has_outer_loop) {
+    const auto loop_entries = loop_info->m_entry_exprs;
+    const auto loop_exits = loop_info->m_exit_exprs;
+    const auto work_amount = loop_info->m_work_amount;
+    const auto work_amount_increment = loop_info->m_increment;
+
+    OutputVector loop_in_outputs, loop_out_outputs;
+    std::vector<LoweredExprPtr> loop_in_exprs, loop_out_exprs;
+    get_io_exprs(linear_ir, loop_entries, loop_exits,
+                 loop_in_exprs, loop_out_exprs,
+                 loop_in_outputs, loop_out_outputs);
+
+    // If we don't need Loop (explicit execution without cycles), we don't explicitly insert the Loop expressions
+    const auto subtensor_in = loop_in_exprs.empty() ? std::vector<size_t>{} : loop_in_exprs.front()->get_inputs().front()->get_subtensor();
+    const bool explicit_execution = (work_amount == 0) || (dim_idx == 0 && subtensor_in.size() > 1 && subtensor_in.back() == work_amount);
+    if (explicit_execution) {
```

Comment:
Ok, then do we still need tail optimizations and `evaluate_once` flag? Or we can handle all such cases here?